### PR TITLE
PR Support ddo credentials in download 170

### LINKED ISSUE #170: Support ddo credentials in download
There is no code to obey DDO credentials section.
Let's implement it

### PR DESCRIPTION
**Pull request by an7e**
Fixes #170

Changes proposed in this PR:

- added check credentials to allow/deny access
- additional validation of ddo

### CODE REVIEW

File: src/components/core/downloadHandler.ts [hunk 1]
```diff
@@ -53,7 +53,65 @@ export async function handleDownload(
...
+      stream: null,
+      status: {
+        httpStatus: 500
+      },
+      error: 'Error: DDO malformed or disabled'
+    }
+  }
+
+  // check credentials
+  if (ddo.credentials) {
```
Comment (alexcos20) [inline src/components/core/downloadHandler.ts:69]: please create a function, like all other checks

File: src/components/core/downloadHandler.ts [hunk 2]
```diff
@@ -53,7 +53,65 @@ export async function handleDownload(
...
+      type: 'address',
+      values: [task.consumerAddress]
+    }
+
+    // check allow access
+    if (Array.isArray(ddo.credentials.allow) && ddo.credentials.allow.length > 0) {
+      const allowCredential = ddo.credentials.allow.find(
+        (credential) =>
+          credential.type === consumerCredentials.type &&
+          credential.values.includes(consumerCredentials.values[0])
```
Comment (alexcos20) [inline src/components/core/downloadHandler.ts:80]: how do you handle case mismatches ?

File: src/components/core/downloadHandler.ts [hunk 3]
```diff
@@ -53,7 +53,65 @@ export async function handleDownload(
...
+          stream: null,
+          status: {
+            httpStatus: 500
+          },
+          error: `Error: Access to asset ${ddo.id} was denied`
+        }
+      }
+    }
+  }
+
```
Comment (alexcos20) [inline src/components/core/downloadHandler.ts:113]: where are the tests?

**Review by alexcos20 — CHANGES_REQUESTED** [review]
this is not complete

**Comment by alexcos20** [comment]
and tests are failing

File: src/components/core/downloadHandler.ts (hunk 3 above)
Comment (an7e) [reply src/components/core/downloadHandler.ts:83]: @alexcos20 There are two tests here, one implicit - the bootstrap test has the consumer address specified in DDO as allowed, and another one explicit - test should not allow to download the asset with different consumer address

File: src/utils/credentials.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+  // check allow access
+  if (Array.isArray(credentials.allow) && credentials.allow.length > 0) {
+    const allowCredential = findCredential(credentials.allow, consumerCredentials)
+    if (!allowCredential) {
+      return false
+    }
+  }
+
+  // check deny access
```
Comment (alexcos20) [inline src/utils/credentials.ts:36]: deny should always come first in checks

File: tests/integration/download.test.ts [hunk 5]
```diff
@@ -296,4 +296,28 @@ describe('Download Tests', () => {
     assert(response.status.httpStatus === 200, 'http status not 200')
     expect(response.stream).to.be.instanceOf(Readable)
   })
+
+  it('should not allow to download the asset with different consumer address', async function () {
```
Comment (alexcos20) [inline tests/integration/download.test.ts:300]: can you create unit tests for checkCredentials with different scenarios?

**Review by alexcos20 — CHANGES_REQUESTED** [review]

**Review by alexcos20 — APPROVED** [review]

**Review by bogdanfazakas — APPROVED** [review]
lgtm

**Pull request merged**